typings/attach.d.ts

```diff
@@ -16,15 +16,23 @@ export interface IAttachOptions {
   /**
    * Whether input should be written to the backend. Defaults to `true`.
    */
-  bidirectional?: boolean,
+  bidirectional?: boolean;
   
   /**
    * Whether to use UTF8 binary transport for incoming messages. Defaults to `false`.
    * Note: This must be in line with the server side of the websocket.
    *       Always send string messages from the backend if this options is false,
    *       otherwise always binary UTF8 data.
    */
-  inputUtf8?: boolean
+  inputUtf8?: boolean;
+
+  /**
+   * Whether to send a '#READY#' message upon activation.
+   * If set the server shall not send data before it has seen this message.
+   * Always set this to avoid losing the first messages.
+   * Since this options needs additional preparations on server side it defaults to `false`.
+   */
+  sendReady?: boolean;
```

Comment:
Is this common practice when dealing with websockets or something? I would have thought this would be handled by the connection stuff, any links that explain the problem in more detail?

Reply:
Well the problem is on our side or how the attach addon sets up the terminal connection  with a websocket object:
- user creates websocket: connection gets established and upgraded in parallel (in browser C++ realms)
- attach addon ctor is called with that instance some times later: by that time the actual state may be still `CONNECTING` or already `OPEN` (or `CLOSING/CLOSED` upon errors)
- if the websocket can be established very fast, the demo server starts sending chunks right away, but the attach addon might not yet have registered the terminal callbacks - lost messages

I see this behavior in the puppeteer tests with this code:
```TS
const ws = new WebSocket('ws://localhost:${port}');
const attachAddon = new window.AttachAddon(ws);  // to late to see all messages
window.term.loadAddon(attachAddon);  // to late
```

The problem is the time span from `new WebSocket` to `loadAddon` with a server not waiting for the latter action, thus sending data right away while no client listener is attached yet. With a READY message a server can explicitly wait for that. Its like an additional ACK telling the server "yes I am ready to consume pty chunks".

I was not able to find reliable information about this particular behavior. The best I found was [this](https://stackoverflow.com/a/34900879), but the marked answer is kinda wrong since it does not reflect the behavior I see with the puppeteer tests (it shows the behavior of answer 2). Also the parts he is referring to are not in the active [spec](https://html.spec.whatwg.org/multipage/web-sockets.html#handler-websocket-onmessage) anymore.

Edit: Reading answer 2 with its comments more closely it actually might be a result of `addEventListener` vs. `onmessage`. Have to test this.